molvis/molvispython/python/processors/MolecularStructureSource.py

```diff
@@ -0,0 +1,199 @@
+
+import molvisdata
+
+from Bio.PDB.MMCIFParser import MMCIFParser
+from Bio.PDB.PDBParser import PDBParser
+from Bio.PDB import CaPPBuilder
+
+import numpy as np
+
+
+class MolecularStructureSource(ivw.Processor):
+    def __init__(self, id, name):
+        ivw.Processor.__init__(self, id, name)
+        
+        self.addOutport(ivw.data.MeshOutport("mesh"))
+        self.addOutport(ivwmolvis.MolecularStructureOutport("molecule"))
+
+        self.filename = ivw.properties.FileProperty("filename", "Structure Filename", "", "molecularstructure")
+        self.filename.addNameFilter(ivw.properties.FileExtension("cif", "mmCIF file"))
+        self.filename.addNameFilter(ivw.properties.FileExtension("pdb", "PDB file"))
+        self.addProperty(self.filename, owner=False)
+
+        self.dataset = ivw.properties.StringProperty("dataset", "Data")
+        self.addProperty(self.dataset, owner=False)
+        self.dataset.readOnly = True
+
+    @staticmethod
+    def processorInfo():
+        return ivw.ProcessorInfo(
+            classIdentifier = "org.inviwo.MolecularStructureSource", 
+            displayName = "Molecular Structure Source",
+            category = "MolVis",
+            codeState = ivw.CodeState.Stable,
+            tags = ivw.Tags.PY
+        )
+
+    def getProcessorInfo(self):
+        return MolecularStructureSource.processorInfo()
+
+    def initializeResources(self):
+        pass
+
+    def process(self):
+        if not self.filename.value or not os.path.isfile(self.filename.value):
+            return
+
+        self.dataset.value = self.filename.value.split("/")[-1]
+
+        mesh, molstruct = self.parseCIF(self.filename.value)
+        self.outports.mesh.setData(mesh)
+        self.outports.molecule.setData(molstruct)
+
+    # @staticmethod
+    # def extractAtomName(atom):
+    #     return molvisdata.extractAtomName(atom.get_name())
+
+    @staticmethod
+    def parseCIF(filename):
+        mesh = ivw.data.Mesh()
+
+        ext = filename.split(".")[-1]
+        if ext == 'gz' or ext == 'bz2':
+            ext = filename.split(".")[-2]
+        print(ext)
+
+        if (ext.startswith('pdb')):
+            print('pdb...')
+            parser = PDBParser(PERMISSIVE=1)
+        elif (ext.startswith('cif')):
+            # initialize mmcif parser
+            parser = MMCIFParser()
+        else:
+            ivw.LogError("unsupported extension '{}'".format(ext))
+            return mesh
+
+        structureName = filename.split("/")[-1]
+        structure = parser.get_structure(structureName, filename)
+
+        pos = []
+        bfactors = []
+        #structureId = []
+        modelId = []
+        chainId = []
+        residueId = []
+        atomFullName = []
+
+        modelDict = dict()
+        chainDict = dict()
+        residueDict = dict()
+
+        def addToDict(d, key):
+            if key not in d:
+                d[key] = len(d)
+            return d[key]
+
+        """def get_full_id(self): 
```

Comment:
remove?

Reply:
This is for reference to BioPython and the tuples.